File: app/models/attendance.py
import datetime
from google.cloud.firestore import GeoPoint


class Attendance:
    def __init__(
        self,
        date: datetime.date,
        user_id: str,
        user_name: str,
        check_in_at: datetime.datetime,
        commitment_time: str,
        ip_address: str,
        lat_lng: GeoPoint,
        place_name: str,
        time_differece: str,
    ):
        self.date = date
        self.user_id = user_id
        self.user_name = user_name
        self.check_in_at = check_in_at
        self.commitment_time = commitment_time
        self.ip_address = ip_address
        self.lat_lng = lat_lng
        self.place_name = place_name
        self.time_differece = time_differece

    def __repr__(self):
        return f"<Attendance(user_id={self.user_id}, user_name={self.user_name}, check_in_at={self.check_in_at}, commitment_time={self.commitment_time}, ip_address={self.ip_address}, lat_lng={self.lat_lng}, place_name={self.place_name})>"

File: frontend/lib/repository/attendance.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import 'package:intl/intl.dart';

import '../models/attendance.dart';

final attendanceRepositoryProvider = Provider((ref) => AttendanceRepository());

class AttendanceRepository {
  final _ref = FirebaseFirestore.instance.collection("attendances");

  Future<Attendance?> getAttendance(String userId, DateTime date) async {
    final dateString = DateFormat("yyyy-MM-dd").format(date);

    final doc = await _ref.doc(dateString).get();
    if (!doc.exists) {
      return null;
    }
    final dataMap = doc.data() as Map<String, dynamic>;

    for (final entry in dataMap.entries) {
      if (entry.key == userId) {
        final data = entry.value as Map<String, dynamic>;
        return Attendance(
          userId: userId,
          userName: data["userName"],
          date: date,
          checkInAt: (data["checkInAt"] as Timestamp).toDate(),
          commitmentTime: data["commitmentTime"],
          ipAddress: data["ipAddress"],
          latLng: data["latLng"],
          placeName: data["placeName"],
          timeDifferenceSeconds: data["timeDifferenceSeconds"],
        );
      }
    }

    return null;
  }

  Future<List<Attendance>> getUserAttendances(
      String userId, List<DateTime> dates) async {
    List<Attendance> userAttendances = [];

    for (final date in dates) {
      final attendance = await getAttendance(userId, date);
      if (attendance != null) {
        userAttendances.add(attendance);
      }
    }

    debugPrint("Get ${userAttendances.length} attendances");

    return userAttendances;
  }
}
